Update cookbook for manually setting span context. (#1262)
open-telemetry/opentelemetry.io: content/en/docs/instrumentation/python/cookbook.md

## Manually setting span context

Usually your application or serving framework will take care of propagating your trace context for you. But in some cases, you may need to save your trace context (with `.inject`) and restore it elsewhere (with `.extract`) yourself.

```python
from opentelemetry import trace, context
from opentelemetry.trace import NonRecordingSpan, SpanContext, TraceFlags
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import ConsoleSpanExporter, BatchSpanProcessor
from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator

# Set up a simple processor to write spans out to the console so we can see what's happening.
trace.set_tracer_provider(TracerProvider())
trace.get_tracer_provider().add_span_processor(BatchSpanProcessor(ConsoleSpanExporter()))

tracer = trace.get_tracer(__name__)

> # A TextMapPropagator can use a regular Python dict as its Carrier.

Reviewer: ```suggestion
# A TextMapPropagator works with any dict-like object as its Carrier by default. You can also implement custom getters and setters.
```

> carrier = {}
> with tracer.start_as_current_span('first-trace'):
>     # Write the current context into the carrier.
>     TraceContextTextMapPropagator().extract(carrier)

Reviewer: I think you meant `inject()` here

```suggestion
with tracer.start_as_current_span('first-trace'):
    carrier = {}
    # Write the current context into the carrier.
    TraceContextTextMapPropagator().inject(carrier)
```

# The below might be in a different thread, on a different machine, etc.
# As a typical example, it would be on a different microservice and the carrier would
# have been forwarded via HTTP headers.

# Extract the trace context from the carrier.

> # Here's what a typical carrier might look like, as it would have been extracted above.

Reviewer: ```suggestion
# Here's what a typical carrier might look like, as it would have been injected above.
```

carrier = {'traceparent': '00-a9c3b99a95cc045e573e163c3ac80a77-d99d251a8caecd06-01'}

> # Then we use a propagator to get a trace context from it.

Reviewer: nit span context is a subset of the overall context

```suggestion
# Then we use a propagator to get a context from it.
```

)
ctx = trace.set_span_in_context(NonRecordingSpan(span_context))

# Now there are a few ways to make use of the trace context.

# You can pass the context object when starting a span.
with tracer.start_as_current_span('child', context=ctx) as span:
    span.set_attribute('primes', [2, 3, 5, 7])

# Or you can make it the current context, and then the next span will pick it up.

> context.attach(ctx)

Reviewer: The best practice here would be wrapping this in a `try` and detaching in the `finally`. I believe it would look like

```py
try:
  token = context.attach(ctx)
  ...
finally:
  context.detach(token)
```

otherwise the context can leak

Author: Thanks, updated. I put the `attach` before the `try` since if that fails you won't have a `token` to detach with.
